src/cursor/cur_std.c

```diff
@@ -849,67 +865,199 @@ __wt_cursor_get_hash(
+    WT_RET(__wt_config_gets_def(session, cfg, "checkpoint", 0, &cval));
+    if (cval.val) {
+return_false:
+        *cacheablep = false;
+    } else {
+        *cacheablep = true;
+    }
+
+    return (0);
+}
+
+/*
+ * __cursor_reuse_or_init --
+ *     Initialization shared between reuse of a cached cursor and initialization of a new cursor.
+ *
+ * Flags set or cleared by this function: WT_CURSTD_APPEND, WT_CURSTD_DEBUG_RESET_EVICT.
+ *
+ * Flags set by this function: WT_CURSTD_RAW.
+ *
+ * Flags cleared by this functions: WT_CURSTD_CACHEABLE.
+ */
+static int
+__cursor_reuse_or_init(WT_SESSION_IMPL *session, WT_CURSOR *cursor, const char *cfg[],
+  bool *readonlyp, WT_CURSOR **ownerp, WT_CURSOR **cdumpp)
+{
+    WT_CONFIG_ITEM cval;
+    WT_CURSOR *cdump;
+    WT_CURSOR *owner;
+
+    if (cfg != NULL) {
+        /*
+         * WT_CURSTD_APPEND is only relevant to column stores.
+         */
+        if (WT_CURSOR_RECNO(cursor)) {
+            WT_RET(__wt_config_gets_def(session, cfg, "append", 0, &cval));
+            if (cval.val != 0)
+                F_SET(cursor, WT_CURSTD_APPEND);
+            else
+                F_CLR(cursor, WT_CURSTD_APPEND);
+        }
+
+        /* WT_CURSTD_RAW */
+        WT_RET(__wt_config_gets_def(session, cfg, "raw", 0, &cval));
+        if (cval.val != 0)
+            F_SET(cursor, WT_CURSTD_RAW);
+
+        /* WT_CURSTD_DEBUG_RESET_EVICT */
+        WT_RET(__cursor_config_debug(cursor, cfg));
+    }
+
+    /* Readonly? */
+    WT_RET(__any_cursor_must_be_readonly(session, cfg, readonlyp));
+
+    /* WT_CURSTD_CACHEABLE */
+    if (*readonlyp) /* We do not cache read-only cursors. */
+        F_CLR(cursor, WT_CURSTD_CACHEABLE);
+
+    /*
+     * dump If an index cursor is opened with dump, then this function is called on the index files,
+     * with the dump config string, and with the index cursor as an owner. We don't want to create a
+     * dump cursor in that case, because we'll create the dump cursor on the index cursor itself.
+     */
+    WT_RET(__wt_config_gets_def(session, cfg, "dump", 0, &cval));
+    owner = ownerp ? *ownerp : NULL;
+    if (cval.len != 0 && owner == NULL) {
+        uint64_t dump_flag;
+        if (WT_STRING_MATCH("json", cval.str, cval.len))
+            dump_flag = WT_CURSTD_DUMP_JSON;
+        else if (WT_STRING_MATCH("print", cval.str, cval.len))
+            dump_flag = WT_CURSTD_DUMP_PRINT;
+        else if (WT_STRING_MATCH("pretty", cval.str, cval.len))
+            dump_flag = WT_CURSTD_DUMP_PRETTY;
+        else if (WT_STRING_MATCH("pretty_hex", cval.str, cval.len))
+            dump_flag = WT_CURSTD_DUMP_HEX | WT_CURSTD_DUMP_PRETTY;
+        else
+            dump_flag = WT_CURSTD_DUMP_HEX;
+        F_SET(cursor, dump_flag);
+
+        /*
+         * Dump cursors should not have owners: only the top-level cursor should be wrapped in a
+         * dump cursor.
+         */
+        WT_RET(__wt_curdump_create(cursor, owner, &cdump));
```

Comment:
I don't think we should be creating new cursors inside a config flag init function. Ideally if possible to move this to the caller would be better.

Reply:
Not possible.

Moving it to the caller would duplicate this code. It is called from two places: __wt_cursor_cache_get() and __wt_cursor_init().

This function is for any cursor initialization shared between __wt_cursor_cache_get() and __wt_cursor_init().

Creating a cursor is due to how dump cursors were implemented. Someone explained dump cursors were implemented as wrapping the original cursor instead of adding a characteristic to the original cursor. The reason is performance. Non-dump cursors do not need to check a dump characteristic.